pkg/mutator/handler.go

```diff
@@ -15,22 +15,114 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-admission-controller/pkg/filter"
+	"github.com/giantswarm/azure-admission-controller/pkg/generic"
 )
 
 type Mutator interface {
-	Log(keyVals ...interface{})
-	Mutate(ctx context.Context, review *v1beta1.AdmissionRequest) ([]PatchOperation, error)
+	generic.Decoder
+	generic.Logger
 	Resource() string
 }
 
+type CreateMutator interface {
+	Mutator
+	Mutate(ctx context.Context, object interface{}) ([]PatchOperation, error)
+}
+
+type UpdateMutator interface {
+	Mutator
+	MutateUpdate(ctx context.Context, oldObject interface{}, object interface{}) ([]PatchOperation, error)
+}
+
 var (
 	scheme        = runtime.NewScheme()
 	codecs        = serializer.NewCodecFactory(scheme)
 	Deserializer  = codecs.UniversalDeserializer()
 	InternalError = errors.New("internal admission controller error")
 )
 
-func Handler(mutator Mutator) http.HandlerFunc {
+type HandlerFactoryConfig struct {
+	CtrlClient client.Client
+}
+
+type HandlerFactory struct {
+	ctrlClient client.Client
+}
+
+func NewHandlerFactory(config HandlerFactoryConfig) (*HandlerFactory, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+
+	h := &HandlerFactory{
+		ctrlClient: config.CtrlClient,
+	}
+
+	return h, nil
+}
+
+func (h *HandlerFactory) NewCreateHandler(mutator CreateMutator) http.HandlerFunc {
```

Comment:
Separate `NewCreateHandler` and `NewUpdateHandler` functions, because due to the second one having to decode one more object, and they call different funcs - `Mutate` and `MutateUpdate` respectively.